scripts/upgrade.sh

```diff
@@ -0,0 +1,191 @@
+            exit 1 ;;
+    esac
+done
+
+TMPFILES=$(mktemp -d)
+
+cat << EOF >> "${LOGFILE}"
+--- MetalK8s Upgrade started on $(date -u -R) ---
+EOF
+
+exec > >(tee -ia "${LOGFILE}") 2>&1
+
+cleanup() {
+    rm -rf "${TMPFILES}" || true
+}
+
+trap cleanup EXIT
+
+
+run_quiet() {
+    local name=$1
+    shift 1
+
+    echo -n "> ${name}..."
+    local start
+    start=$(date +%s)
+    set +e
+    "$@" 2>&1 | tee -ia "${LOGFILE}" > "${TMPFILES}/out"
+    local RC=$?
+    set -e
+    local end
+    end=$(date +%s)
+
+    local duration=$(( end - start ))
+
+    if [ $RC -eq 0 ]; then
+        echo " done [${duration}s]"
+    else
+        echo " fail [${duration}s]"
+        cat >/dev/stderr << EOM
+
+Failure while running step '${name}'
+
+Command: $@
+
+Output:
+
+<< BEGIN >>
+EOM
+        cat "${TMPFILES}/out" > /dev/stderr
+
+        cat >/dev/stderr << EOM
+<< END >>
+
+This script will now exit
+
+EOM
+
+        exit 1
+    fi
+}
+
+run_verbose() {
+    local name=$1
+    shift 1
+
+    echo "> ${name}..."
+    "$@"
+}
+
+run() {
+    if [ "$VERBOSE" -eq 1 ]; then
+        run_verbose "${@}"
+    else
+        run_quiet "${@}"
+    fi
+}
+
+die() {
+    echo 1>&2 "$@"
+    return 1
+}
+
+get_salt_container() {
+    local -r max_retries=10
+    local salt_container='' attempts=0
+
+    while [ -z "$salt_container" ] && [ $attempts -lt $max_retries ]; do
+        salt_container="$(crictl ps -q \
+            --label io.kubernetes.pod.namespace=kube-system \
+            --label io.kubernetes.container.name=salt-master \
+            --state Running)"
+        (( attempts++ ))
+    done
+
+    if [ -z "$salt_container" ]; then
+        echo "Failed to find a running 'salt-master' container" >&2
+        exit 1
+    fi
+
+    echo "$salt_container"
+}
+
+__orchestrate_upgrade() {
+    run "Performing bootstrap node upgrade (this may take a while)" \
+    "${SALT_CALL}" --local state.sls metalk8s.roles.bootstrap saltenv=metalk8s-"$DESTINATION_VERSION" \
+        pillar="{'metalk8s': {'endpoints': $(salt-call --out txt pillar.get metalk8s:endpoints \
+        | cut -c 8-)}}"
+
+    SALT_MASTER_CALL=(crictl exec -i "$(get_salt_container)")
+
+    run "Launching rolling upgrade (this may take a while)" \
+    "${SALT_MASTER_CALL[@]}" salt-run state.orchestrate metalk8s.orchestrate.upgrade saltenv=metalk8s-"$DESTINATION_VERSION" \
+        pillar="{'orchestrate': {'dest_version': '$DESTINATION_VERSION'}}"
+}
+
+__orchestrate_precheck() {
+    if [ -z "$DESTINATION_VERSION" ]; then
+        die "No destination version given"
+    fi
+    SALT_MASTER_CALL=(crictl exec -i "$(get_salt_container)")
+
+    run "Performing Pre-Upgrade checks" \
+    "${SALT_MASTER_CALL[@]}" salt-run state.orchestrate metalk8s.orchestrate.upgrade.precheck saltenv=metalk8s-"$DESTINATION_VERSION" \
+        pillar="{'orchestrate': {'dest_version': '$DESTINATION_VERSION'}}"
+}
+
+if [ $DRY_RUN -eq 1 ]; then
+    __orchestrate_precheck
+    else
+        __orchestrate_upgrade
```

Comment:
No I answered this one lets keep the precheck always . Try to do something like in @sayf-eddine-scality PR
```
run "Performing Pre-Downgrade checks" precheck_downgrade
[ $DRY_RUN -eq 1 ] && exit 0
run "Launching the downgrade" launch_downgrade
run "Downgrading bootstrap" downgrade_bootstrap
```
Replacing Downgrade by Upgrade of course.

Also like in @sayf-eddine-scality PR try to use `run` the command here.
Please sync together to have the same "template of script"
I mean both script should be very similar.

Reply:
Well noted